### PR TITLE
Graph Compiler For Multi-stream

### PR DESCRIPTION
**Pull request by alpha0422**
This PR contains an experimental graph compiler on top of `torch.compile`. We used it for a couple of studies. In this PR we'd like to expose one functionality, which is scheduling independent operators in the graph on multi-streams automatically.

Right now the best use case of it is for small workloads where a single GEMM/Conv can’t fully utilize all the GPU resources, especially during back-propagation.

We achieved 5%-9% E2E performance imporvement for Stable Diffusion with small batch size and full iteration CUDA graph.                                                                       
                                                                                    
On importing `apex.contrib.torchsched`, a new `torch.compile` backend `torchsched` is registered:
                                                                                    
```python                                                                           
import apex.contrib.torchsched                                                                   
                                                                                    
compiled = torch.compile(model, backend="torchsched")                               
```                                                                                 
                                                                                    
Alternatively, one can import `torchsched` at first and then set `torchsched` as    
default backend:                                                                    
                                                                                    
```python                                                                           
import apex.contrib.torchsched                                                                   
apex.contrib.torchsched.set_default_backend("torchsched")                                        
                                                                                    
compiled = torch.compile(model)                                                     
```

### CODE REVIEW

File: apex/contrib/torchsched/inductor/event.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        """Wait for this event to complete by a given stream.
+
+        Args:
+            stream_idx: The index of the stream that will be waiting for this event to complete.
+
+        Returns:
+            An internal data structure that depicts stream <-> event dependency.
+
+        Note:
+            This method doesn't directly generate a event waiting in the final program. Instead
```
Comment (eqy) [inline apex/contrib/torchsched/inductor/event.py:101]: The wording here is confusing, is "necessarily" a more accurate description rather than "directly?"

File: apex/contrib/torchsched/inductor/graph.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    Args:
+        patch: Whether to patch Inductor `GraphLowering` with multi-stream
+            scheduler. Set to `False` to restore the default `torch.compile`
+            behavior. (default: `True`)
+    """
+    if patch:
+        GraphLowering.codegen = _codegen
+    else:
+        GraphLowering.codegen = _origin_codegen
```
Comment (eqy) [inline apex/contrib/torchsched/inductor/graph.py:59]: Just checking, are there no more convenient registration (e.g., custom backend) mechanisms in torch.compile to support this? Or is the only way to overwrite the codegen pass?

File: apex/contrib/torchsched/inductor/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+                if len(critical_path_per_depth[depth]) < len(path):
+                    critical_path_per_depth[depth] = path
+                for user in node_users[node]:
+                    visit(user, depth + 1, path)
+
+        graph_entries = [n for n in self.nodes if not n.unmet_dependencies]
+        for entry in graph_entries:
+            visit(entry, depth=1, prev=set())
+
+        max_depth, longest_critical_path = sorted(critical_path_per_depth.items(), reverse=True)[0]
```
Comment (eqy) [inline apex/contrib/torchsched/inductor/scheduler.py:157]: is there an estimate from how far from optimal this strategy is compared to e.g., flop count, measured execution time, etc.?

File: apex/contrib/torchsched/ops/layer_norm.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    Returns:
+        CuDNNManager: Global CuDNN manager.
+    """
+    global _global_cudnn_manager
+    if _global_cudnn_manager is None:
+        _global_cudnn_manager = CuDNNManager()
+    return _global_cudnn_manager
+
+
+class LayerNormGraphFactory:
```
Comment (eqy) [inline apex/contrib/torchsched/ops/layer_norm.py:81]: Have you evaluated cuDNN layernorm performance vs. native PyTorch layer norm? In my testing I've found that in many cases, especially for smaller shapes native PyTorch is faster.

File: apex/contrib/torchsched/utils.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+cudart = torch.cuda.cudart()
+
+
+def partition_weighted(
+    weights: list[int] | tuple[int, ...],
+    sm_count: int | None = None,
+    rounding_multiple: int = 1,
+) -> list[int]:
+    """Partition the SMs based on their weights, rounding to a specified multiple.
```
Comment (eqy) [inline apex/contrib/torchsched/utils.py:21]: unclear what "weights" are here, are these GEMM weights?

File: apex/contrib/torchsched/utils.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+    This function distributes a total number of SMs proportionally to their weight counts,
```
Comment (eqy) [inline apex/contrib/torchsched/utils.py:23]: "distributes a total number of SMs" is somewhat awkward, is the meaning "map #GEMM weights to a number of SMs?"

File: apex/contrib/torchsched/utils.py (hunk 5 above)
Comment (eqy) [reply apex/contrib/torchsched/utils.py:21]: maybe illustrate what the relevant use-case is---are these GEMMs expected to be concurrent (and that's why the sum should be the given `sm_count`?

**Review by eqy — COMMENTED** [review]
left some initial comments mostly trying to understand doc comments

File: apex/contrib/torchsched/__init__.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    if backend is None:
+        backend = _DEFAULT_BACKEND
+    return __torch_compile__(*args, backend=backend, **kwargs)
+
+
+# Monkey patch torch.compile to set default backend
+torch.compile = torchsched_compile
+
+__all__ = ["get_backend", "set_default_backend"]
```
Comment (crcrpar) [inline apex/contrib/torchsched/__init__.py:79]: could you move `__all__` to right after imports?

File: apex/contrib/torchsched/inductor/scheduler.py (hunk 3 above)
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/scheduler.py:157]: No, we don't have that yet. We need a strategy to achieve it. Some other studies in the project do use measured execution time, but not for multi-stream scheduling.

File: apex/contrib/torchsched/ops/layer_norm.py (hunk 4 above)
Comment (alpha0422) [reply apex/contrib/torchsched/ops/layer_norm.py:81]: That's interesting, native layer norm perfoms quite worse for our cases, especially on B200.

File: apex/contrib/torchsched/config.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+enable_post_grad_pre_pass = False
+
+# Number of CUDA streams used for multi-stream scheduling.
+# The first stream will be critical path stream, operators on non-critical path will be
+# scheduled to other streams in a round-robin way.
+num_streams = int(os.getenv("TORCH_SCHED_NUM_STREAMS", "8"))
+
+if TYPE_CHECKING:
+    from torch.utils._config_typing import *  # noqa: F401, F403
+
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/config.py:26]: In the past the code can't pass pre-commit check unless adding those lines. I just checked we don't need it anymore.

File: apex/contrib/torchsched/inductor/event.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    increasing indices as compilation goes along. It also maintains a pool of materialized cuda
+    events that symbolic events can reuse.
+    """
+
+    def __init__(self) -> None:
+        """Initialize a event factory."""
+        self.symbolic_event_idx: itertools.count = itertools.count(start=1)
+        self.materialized_event_idx: itertools.count = itertools.count(start=1)
+        self.available_materialized_events: set[str] = set()
+        self._entrance_event: CudaEventSym | None = None
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/event.py:155]: Hmm, this class is not pure data, it contains a few other logic. Maybe normal class is better?

File: apex/contrib/torchsched/utils.py (hunk 5 above)
Comment (alpha0422) [reply apex/contrib/torchsched/utils.py]: Removed this file since multi-stream doesn't use it.

File: apex/contrib/torchsched/utils.py (hunk 6 above)
Comment (alpha0422) [reply apex/contrib/torchsched/utils.py]: Removed the file, we can discuss online.

File: apex/contrib/torchsched/__init__.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+import torch._inductor
+from torch._dynamo import list_backends
+from torch._dynamo import register_backend
+from torch._inductor.compile_fx import compile_fx_inner
+
+from .backend import get_backend
+
+if TYPE_CHECKING:
+    from torch._ops import OpOverload
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/__init__.py]: Thanks, changed.

File: apex/contrib/torchsched/inductor/__init__.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+"""Scheduling abstractions on PyTorch Inductor level."""
+
+from __future__ import annotations
+
+from apex.contrib.torchsched.inductor.graph import patch_graph_lowering
+
+__all__ = ("patch_graph_lowering",)
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/__init__.py]: Fixed.

File: apex/contrib/torchsched/inductor/__init__.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+"""Scheduling abstractions on PyTorch Inductor level."""
+
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/__init__.py]: Removed.

File: apex/contrib/torchsched/inductor/scheduler.py [hunk 13]
```diff
@@ -0,0 +1,537 @@
+"""Scheduling abstractions on PyTorch Inductor Scheduler level."""
+
+from __future__ import annotations
+
+import collections
+import copy
+import itertools
+import re
+import typing
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/scheduler.py]: Changed, thanks.

File: apex/contrib/torchsched/inductor/scheduler.py [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+from torch._inductor.utils import device_need_guard
+from torch._inductor.virtualized import V
+
+import apex.contrib.torchsched.config as config
+from apex.contrib.torchsched.inductor.event import CudaEventFactory
+from apex.contrib.torchsched.inductor.event import CudaEventSym
+from apex.contrib.torchsched.inductor.wrapper import DEFAULT_STREAM_IDX
+from apex.contrib.torchsched.inductor.wrapper import EnterCudaStreamContextLine
+
+if typing.TYPE_CHECKING:
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/inductor/scheduler.py]: Changed, thanks.

File: apex/contrib/torchsched/ops/__init__.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
+"""Custom PyTorch operators."""
+
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/torchsched/ops/__init__.py]: Changed, thanks.

File: apex/contrib/torchsched/__init__.py (hunk 7 above)
Comment (alpha0422) [reply apex/contrib/torchsched/__init__.py]: Done.

**Review by crcrpar — APPROVED** [review]

**Review by eqy — APPROVED** [review]
